webapp/src/main/java/org/mediacloud/cliff/servlet/service/ParseTextService.java

```diff
@@ -0,0 +1,166 @@
+        executorService.shutdown();
+        return BatchRawParseTextResponse.builder()
+                .results(responses)
+                .actualMilliseconds(stopWatch.getTime())
+                .sumOfMilliseconds(milliseconds)
+                .build();
+    }
+
+    private static Map<Pair<String, String>, Long> toCountryStateCountMap(List<Mention> mentions) {
+        return mentions.stream()
+                .map(m -> Pair.of(m.getCountryCode(), m.getStateCode()))
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    }
+
+    private static Map<String, Long> toCountryMap(List<Mention> mentions) {
+        return mentions.stream()
+                .map(Mention::getCountryCode)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    }
+
+    private static Optional<String> getCountyName(Mention mention) {
+        HashMap<?, ?> geonameInfo = (HashMap<?, ?>) ParseManager.getGeoNameInfo(mention.getId()).get("results");
+
+        while (geonameInfo.containsKey("parent")) {
+            if (COUNTY_FEATURE_CODE.equals(geonameInfo.get("featureCode"))) {
+                return Optional.of((String) geonameInfo.get("name"));
+            }
+            geonameInfo = (HashMap<?, ?>) geonameInfo.get("parent");
+        }
+        
+        return Optional.empty();
+    }
+
+    private static ParseTextResponse collapseMentions(ParseTextResponse parseTextResponse) {
+        Map<Pair<String, String>, Long> countryStateCountMap = toCountryStateCountMap(parseTextResponse.getMentions());
+        Map<String, Long> countryCountMap = toCountryMap(parseTextResponse.getMentions());
+
+        // Getting most specific place grouping by country and state
+        List<Mention> condensedMentions = parseTextResponse.getMentions().stream()
+                // Getting most specific place grouping by country and state
+                .filter(mention -> {
+                    if (COUNTRY_FEATURE_CODE.equals(mention.getFeatureCode()) && !countryCountMap.get(mention.getCountryCode()).equals(1L)) {
+                        return false;
+                    }
+
+                    Pair<String, String> countryStateKey = Pair.of(mention.getCountryCode(), mention.getStateCode());
+                    return !STATE_FEATURE_CODE.equals(mention.getFeatureCode()) || countryStateCountMap.get(countryStateKey).equals(1L);
+                })
+                // counties need to have the "county" field populated
+                .peek(mention -> {
+                    if (COUNTY_FEATURE_CODE.equals(mention.getFeatureCode())) {
+                        mention.setCounty(mention.getPlace());
+                    }
+                })
+                // "places" need the county field populated.
+                .peek(mention -> {
+                    if (!POPULATED_AREAS_CLASS.equals(mention.getFeatureClass())) {
+                        getCountyName(mention).ifPresent(mention::setCounty);
```

Comment:
Just to understand, at this point we'll be filling the county **only** if we have it available somewhere else in the response? Or are we calling the endpoint to get it?

In the first case, how will we handle getting a county and a place that's not from that county in the same String?

"I live in Alachua County, but I love to visit New York"

Reply:
we are calling the method to get the geoname information, including the parents. After that, we go up until we find the county feature code.

For your example, I'm getting the following response:

<img width="374" alt="Screenshot 2023-01-25 at 10 30 04 AM" src="https://user-images.githubusercontent.com/29154504/214590029-06f0232d-e557-4f78-8203-e1c9f17e2f66.png">

Which is expected.